rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,388 @@
+
+## Proposal: `splunk_hec` Sinks Indexer Acknowledgement
+
+### User Experience
+
+The `splunk_hec` sinks will automatically integrate with indexer
+acknowledgements if the user has enabled it in their Splunk instance (if
+`ackId`’s are present in HEC responses). If indexer acknowledgement is disabled,
+`splunk_hec` sinks will continue to finalize events based on the HEC response
+status code.
+
+Users can configure the `splunk_hec` sinks with the following indexer
+acknowledgement settings
+
+* `acknowledgements.query_interval` The amount of time to wait between requests
+  to `services/collector/ack`. This defaults to 10 seconds as recommended by
+  Splunk.
+* `acknowledgements.retry_limit` The number of retry requests to
+  `services/collector/ack`. This defaults to 30 which, along with the default
+  `query_interval`, is 5 minutes of retrying as recommend by Splunk.
+
+### Implementation
+
+* [Splunk recommendations for client integration with indexer acknowledgement](https://docs.splunk.com/Documentation/Splunk/8.2.3/Data/AboutHECIDXAck#Indexer_acknowledgment_client_behavior)
+
+We generate a single `guid` to use as a channel ID and include this in HEC
+requests. After submitting events to HEC, we will parse the HTTP response for an
+`ackId`. If no `ackId` is found, we rely on the current behavior of setting
+`EventStatus` based solely on the HTTP status code.
+
+If an `ackId` is found, we store it in a `Arc<Mutex<Map<u64, (u8, Sender)>>>`
+shared with a background tokio task which, for all pending `ackId`’s, will query
+`/services/collector/ack`. The `(u8, Sender)` map value represents the number of
+retries remaining and the send end of a one-shot notification channel.
+
+This background task will query at an interval (configured or default) and with
+a retry limit (configured or default). If we receive `true` for an `ackId`,
+we'll notify an awaiting receiver with a `Status::Delivered`. If we receive
+`false` for an `ackId`, we decrement its remaining retry count. When remaining
+retries is `0`, we notify with `EventStatus::Dropped`.
+
+Back in the response handler, we’ll await the receiver. Once a status has been
+received, we remove the `ackId` from the map and proceed. Below is an example of
+response handler behavior.
+
+```rust
+fn call(&mut self, req: HecRequest) -> Self::Future {
+        let mut http_service = self.batch_service.clone();
+        Box::pin(async move {
+            ...
+
+            // handle response
+            let response = http_service.call(req).await?;
+            // if ack_id is found
+            let (tx, rx) = oneshot::channel::<EventStatus>();
+            self.ack_id_to_status_map.insert(ack_id, (30, tx));
+            let event_status = match rx.await {
+                Ok(EventStatus::Delivered) => EventStatus::Delivered,
+                Ok(_) => EventStatus::Dropped,
+                Err(_) => EventStatus::Failed,
+            }
+            self.ack_id_to_status_map.remove(ack_id);
```

Comment:
The ack polling process should be doing the remove here. Actually, it will have to, as sending to a oneshot drops it, hence the map will no longer own it. You could play games with an `Option`, but I doubt it's worth it.

Reply:
Thanks, updated in 8661eeb